### PR TITLE
Catch exceptions in scheduled tasks to prevent unintended cancellation

### PR DESCRIPTION
**Pull request by lhotari**
### Motivation

- Prevent the intended cancellation of scheduled tasks in Pulsar code base.
- [ScheduledExecutorService#scheduleAtFixedRate](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html#scheduleAtFixedRate(java.lang.Runnable,long,long,java.util.concurrent.TimeUnit)) (and [scheduleWithFixedDelay](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html#scheduleWithFixedDelay(java.lang.Runnable,long,long,java.util.concurrent.TimeUnit))) won't schedule the next execution if running the task throws an exception. This is explained [in the javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ScheduledExecutorService.html#scheduleAtFixedRate(java.lang.Runnable,long,long,java.util.concurrent.TimeUnit)). This can lead to unintended cancellation of the scheduled task in failure scenarios.

- There are some failures scenarios where it is hard to determine what causes Pulsar broker or client to get into a state where the service is partially working. One such source of issues could be such that a scheduled task silently gets cancelled. This PR will ensure that this won't happen when scheduled tasks are using the provided solution (Runnables.catchingAndLoggingThrowables).

### Additional context

This is a cross-cutting change that is about the correct usage of ScheduleExecutorService API. The problem and solution is explained in this StackOverflow Answer: https://stackoverflow.com/a/24902026.

Bookkeeper client includes a solution called SafeRunnable (in 2 locations: in https://github.com/apache/bookkeeper/blob/master/bookkeeper-common/src/main/java/org/apache/bookkeeper/common/util/SafeRunnable.java and https://github.com/apache/bookkeeper/blob/master/bookkeeper-server/src/main/java/org/apache/bookkeeper/util/SafeRunnable.java). The SafeRunnable.safeRun method is used extensively in many parts of Bookkeeper code. Some parts of Pulsar code base uses SafeRunnable.safeRun to catch and log exceptions. 

The concept of "SafeRunnable" is confusing. Instead of copying the same solution that Bookkeeper uses, it is intentional that there is no "safe runnable" and the method is called "Runnables.catchingAndLoggingThrowables" to make it explicit that the Runnable is wrapped with handling that catches and logs throwables. There is no "magic".

### Modifications

- Add new static method `org.apache.pulsar.common.util.Runnables.catchingAndLoggingThrowables` to pulsar-common which wraps a `Runnable` with try-catch block to catch and log all Throwables.
- Use this wrapper in most usages of ScheduledExecutorService scheduleAtFixedRate and scheduleWithFixedDelay methods.

### CODE REVIEW

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.util;
+
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class Runnables {
```
Comment (eolivelli) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java:25]: What about adding a private constructor?

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * Wraps a Runnable so that throwables are caught and logged when a Runnable is run.
+     *
+     * The main usecase for this method is to be used in {@link java.util.concurrent.ScheduledExecutorService#scheduleAtFixedRate(Runnable, long, long, TimeUnit)}
+     * calls to ensure that the scheduled task doesn't get cancelled as a result of an uncaught exception.
+     *
+     * @param runnable The runnable to wrap
+     * @return a wrapped Runnable
+     */
+    public static Runnable catchingAndLoggingThrowables(Runnable runnable) {
```
Comment (eolivelli) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java:37]: What about 'safeRun'?
This name sounds too long and verbose to me

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        return () -> {
```
Comment (eolivelli) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java:38]: Lambda usually generate ugly stack traces.
What about using a regular static inner class?

**Review by eolivelli — CHANGES_REQUESTED** [review]
Overall looks good to me

I left some feedback

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java (hunk 2 above)
Comment (lhotari) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java:37]: It's intentionally long and descriptive and does exactly what the method name says. I explained in the PR description why I think SafeRunnable / safeRun is a bad concept.

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java (hunk 1 above)
Comment (lhotari) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java:25]: sure, I'll switch to that solution.

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java (hunk 3 above)
Comment (lhotari) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/Runnables.java:38]: Makes sense, I'll change it to use an ordinary static inner class.

**Review by eolivelli — APPROVED** [review]
LGTM

**Review by merlimat — APPROVED** [review]

**Pull request merged**